File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/ui/toolwindow/pullrequestlist/PullRequestListPresenter.java
package com.duxtinto.intellij.plugin.github.codereviews.ui.toolwindow.pullrequestlist;

import com.duxtinto.intellij.plugin.github.codereviews.di.scopes.ProjectScoped;
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.PullRequestEntity;
import com.intellij.util.ui.ColumnInfo;

import javax.annotation.Nonnull;
import javax.inject.Inject;
import javax.inject.Named;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

@ProjectScoped
public class PullRequestListPresenter implements PullRequestList.Presenter {
    private PullRequestList.View view;
    private final PullRequestList.Model model;

    @Inject
    public PullRequestListPresenter(@Nonnull @Named("default") ColumnInfo<?, ?>[] columnInfos) {
        checkNotNull(columnInfos);

        this.model = new PullRequestListModel(columnInfos);
    }

    @Override
    public void displayPullRequests(List<PullRequestEntity> pullRequests) {
        updateModel(pullRequests);
        if (view != null) {
            view.render(model);
        }
    }

    @Override
    public PullRequestList.Model getModel() {
        return model;
    }

    private void updateModel(List<PullRequestEntity> pullRequests) {
        model.setPullRequests(pullRequests);
    }

    public void setView(@Nonnull PullRequestList.View view) {
        this.view = checkNotNull(view);
    }
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/presentation/toolwindow/pullrequestlist/PullRequestListPresenter.kt
package com.duxtinto.intellij.plugin.github.codereviews.presentation.toolwindow.pullrequestlist

import com.duxtinto.intellij.plugin.github.codereviews.di.scopes.ProjectScoped
import com.duxtinto.intellij.plugin.github.codereviews.domain.pullrequests.PullRequestEntity
import com.intellij.util.ui.ColumnInfo
import javax.inject.Inject
import javax.inject.Named

import com.google.common.base.Preconditions.checkNotNull

@ProjectScoped
class PullRequestListPresenter
    @Inject
    constructor(@Named("default") columnInfos: Array<ColumnInfo<PullRequestEntity, *>>)
    : PullRequestList.Presenter {

    override val model: PullRequestList.Model = PullRequestListModel(columnInfos)
    private var view: PullRequestList.View? = null

    override fun displayPullRequests(pullRequests: List<PullRequestEntity>) {
        updateModel(pullRequests)
        view?.render(model)
    }

    private fun updateModel(pullRequests: List<PullRequestEntity>) {
        model.setPullRequests(pullRequests)
    }

    override fun setView(view: PullRequestList.View) {
        this.view = view
    }
}
